src/cmap/worker.rs

```diff
@@ -191,13 +223,43 @@ impl ConnectionPoolWorker {
             };
 
             match task {
-                PoolTask::CheckOut(result_sender) => self.check_out(result_sender).await,
-                PoolTask::CheckIn(connection) => self.check_in(connection),
-                PoolTask::Clear => self.clear(),
-                PoolTask::HandleConnectionSucceeded(c) => self.handle_connection_succeeded(c),
-                PoolTask::HandleConnectionFailed(error) => self.handle_connection_failed(error),
+                PoolTask::CheckOut(request) => match self.state {
+                    PoolState::Ready => {
+                        self.wait_queue.push_back(request);
+                    }
+                    PoolState::Paused => {
+                        // if receiver doesn't listen to error that's ok.
+                        let _ = request.fulfill(ConnectionRequestResult::PoolCleared);
+                    }
+                },
+                PoolTask::HandleManagementRequest(PoolManagementRequest::CheckIn(connection)) => {
+                    self.check_in(connection)
+                }
+                PoolTask::HandleManagementRequest(PoolManagementRequest::Clear) => self.clear(None),
+                PoolTask::HandleManagementRequest(PoolManagementRequest::MarkAsReady {
+                    completion_handler,
+                }) => {
+                    self.mark_as_ready();
+                    // other end hung up, nothing to do
+                    let _: std::result::Result<_, _> = completion_handler.send(());
+                }
+                PoolTask::HandleManagementRequest(
+                    PoolManagementRequest::HandleConnectionSucceeded(c),
+                ) => self.handle_connection_succeeded(c),
+                PoolTask::HandleManagementRequest(
+                    PoolManagementRequest::HandleConnectionFailed {
+                        error,
+                        error_generation,
+                    },
+                ) => self.handle_connection_failed(error, error_generation),
                 PoolTask::Maintenance => self.perform_maintenance(),
             }
+
+            if self.can_service_connection_request() {
```

Comment:
We still only need to process them after a task is performed, since this worker task has exclusive access to the inner state. If we can't process a connection request at a certain point, we still won't be able to process any more unless another task has been performed that might change that.